File: quiz-nodejs-engine/src/service/QuizUtils.ts
import { QuizSession } from '../models/QuizSession';

export class QuizUtils {
  /**
   * Calculate percentage correct (0-100).
   */
  static calculatePercentage(numCorrect: number, numQuestions: number): number {
    if (numQuestions === 0) return 0;
    return Math.round((numCorrect / numQuestions) * 100 * 100) / 100;
  }

  /**
   * Get a pass/fail grade based on percentage (passing = 70%).
   */
  static getGrade(percentage: number): string {
    if (percentage >= 90) return 'A';
    if (percentage >= 80) return 'B';
    if (percentage >= 70) return 'C';
    if (percentage >= 60) return 'D';
    return 'F';
  }

  /**
   * Determine if a score is passing (>= 70%).
   */
  static isPassing(percentage: number): boolean {
    return percentage >= 70;
  }

  /**
   * Format elapsed time in seconds to a human-readable string.
   */
  static formatDuration(seconds: number): string {
    if (seconds < 60) return `${seconds}s`;
    const mins = Math.floor(seconds / 60);
    const secs = seconds % 60;
    if (mins < 60) return `${mins}m ${secs}s`;
    const hours = Math.floor(mins / 60);
    const remainMins = mins % 60;
    return `${hours}h ${remainMins}m ${secs}s`;
  }

  /**
   * Format a date for display.
   */
  static formatDate(date: Date): string {
    return date.toLocaleString();
  }

  /**
   * Validate that an answer string is A-E.
   */
  static isValidAnswer(answer: string): boolean {
    return /^[A-Ea-e]$/.test(answer);
  }

  /**
   * Build a summary object for a session.
   */
  static buildSessionSummary(session: QuizSession): {
    sessionId: string;
    startedAt: string;
    endedAt: string;
    numQuestions: number;
    numCorrect: number;
    percentage: number;
    grade: string;
    passed: boolean;
    duration: string;
  } {
    const percentage = Number(session.percentageCorrect);
    return {
      sessionId: session.sessionId,
      startedAt: QuizUtils.formatDate(session.startedAt),
      endedAt: session.endedAt ? QuizUtils.formatDate(session.endedAt) : 'In progress',
      numQuestions: session.numQuestions,
      numCorrect: session.numCorrect,
      percentage,
      grade: QuizUtils.getGrade(percentage),
      passed: QuizUtils.isPassing(percentage),
      duration: session.timeTakenSeconds
        ? QuizUtils.formatDuration(session.timeTakenSeconds)
        : 'N/A',
    };
  }
}

File: quiz-engine/quiz-engine-springboot/src/main/java/com/quizengine/util/QuizUtils.java
package com.quizengine.util;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.UUID;

public class QuizUtils {

    private QuizUtils() {}

    public static String generateSessionId() {
        return UUID.randomUUID().toString();
    }

    public static double calculatePercentage(int correct, int total) {
        if (total == 0) return 0.0;
        return Math.round((double) correct / total * 10000.0) / 100.0;
    }

    public static int calculateElapsedSeconds(LocalDateTime start) {
        return (int) Duration.between(start, LocalDateTime.now()).getSeconds();
    }

    public static String formatDuration(int seconds) {
        int hours = seconds / 3600;
        int minutes = (seconds % 3600) / 60;
        int secs = seconds % 60;
        if (hours > 0) {
            return String.format("%dh %dm %ds", hours, minutes, secs);
        } else if (minutes > 0) {
            return String.format("%dm %ds", minutes, secs);
        } else {
            return String.format("%ds", secs);
        }
    }

    public static String gradeResult(double percentage) {
        if (percentage >= 90) return "EXCELLENT";
        if (percentage >= 75) return "PASS";
        if (percentage >= 60) return "BORDERLINE";
        return "FAIL";
    }
}
